### PR TITLE
Move TaintBasedEvictions feature gates to GA

### LINKED ISSUE #87428: [TaintBasedEvictions] Clean up featuregate logic
As part of moving TaintBasedEvictions to GA in 1.18 (https://github.com/kubernetes/kubernetes/issues/87161), we need to clean up the logic around featuregating for this feature. Since it will be GA, it should not require a featuregate anymore.

/kind cleanup
/sig scheduling

### PR DESCRIPTION
**Pull request by Ye-Tian-Zero**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
Move TaintBasedEvictions feature gates to GA
**Which issue(s) this PR fixes**:

Fixes #87428

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
Move TaintBasedEvictions feature gates to GA
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/blob/74e610bb0f7e40862688e8a434c77bfafc53cb9e/keps/sig-scheduling/20200114-taint-based-evictions.md
```

### CODE REVIEW

**Comment by skilxn-go** [comment]
#87161

**Comment by damemi** [comment]
/ok-to-test

Thanks @skilxn-go! @Huang-Wei, are there any other places this needs to be updated?

**Comment by Huang-Wei** [comment]
> are there any other places this needs to be updated?

Here are the things we need to update:

- featuregate to GA and add a comment indicating that the featuregate will be removed in next release (1.19) - please check other graduated features for reference (e.g., TaintNodeByConditions)
- cleanup all logic (including tests) related with this featuregate, and ensure CI is green

@skilxn-go It only resolves #87428. Please remove #87161 in the description section. And we should provide a release-note.

**Comment by skilxn-go** [comment]
Thanks, I'm going to fix it.

**Comment by Huang-Wei** [comment]
@skilxn-go, #82703 is an example for your reference.

File: pkg/controller/nodelifecycle/node_lifecycle_controller.go
Comment (Huang-Wei) [inline pkg/controller/nodelifecycle/node_lifecycle_controller.go]: Can you also remove its definition in type `Controller`?

File: test/integration/scheduler/taint_test.go
Comment (Huang-Wei) [inline test/integration/scheduler/taint_test.go]: You need to also remove the feature gate setting at L645.

**Review by Huang-Wei — COMMENTED** [review]
Thanks @skilxn-go . Some comments below.

**Comment by mattjmcnaughton** [comment]
/remove-sig node

Not sure I'm seeing anything requiring sig-node specific review. Please feel free to add sig-node back if I'm mistaken.

**Comment by skilxn-go** [comment]
@Huang-Wei I fixed them, but still two test failed, seems like flake tests.
/retest

File: test/integration/scheduler/taint_test.go
Comment (haosdent) [inline test/integration/scheduler/taint_test.go]: Need to update

```
staging/src/k8s.io/api/core/v1/well_known_taints.go:    // and feature-gate for TaintBasedEvictions flag is enabled,
staging/src/k8s.io/api/core/v1/well_known_taints.go:    // and feature-gate for TaintBasedEvictions flag is enabled,
```

Comment (skilxn-go) [reply test/integration/scheduler/taint_test.go]: The TaintBasedEvictions feature-gate is node removed now (it'll be removed in v1.19). It's move to ga and locked to true.
So may be it's better to update this comments when we totally remove this feature-gate.
What do you think?

Comment (haosdent) [reply test/integration/scheduler/taint_test.go]: Got it, thx @skilxn-go 's reply!

Comment (skilxn-go) [reply test/integration/scheduler/taint_test.go]: After think about your comment again, I think it's ok to do the clean up work now. So I fixed them, thanks for your review 😄

**Comment by skilxn-go** [comment]
I need some help to determine if this pr need API review. It's auto labeled as kind/api-change after I modify some comments in the last commit. So I think it doesn't need an Api review, but I'm not sure.

**Comment by ahg-g** [comment]
/hold

Before we do that, we need to clearly define who owns the feature and the tests. 

1) The KEP https://github.com/kubernetes/enhancements/pull/1450 lists sig-node as the owning sig, and sig-node and sig-cloud-provider as the approvers. I think the KEP was wrongly merged under sig-scheduling, it should have been placed under keps/sig-node.

2) As discussed in the KEP, before moving the feature to GA, we need to place the integration and e2e tests under the proper sig that owns the feature.

File: test/integration/scheduler/taint_test.go
Comment (Huang-Wei) [reply test/integration/scheduler/taint_test.go]: Can you also update the comments related with `TaintNodesByCondition`? (that feature has been graduated)

**Comment by damemi** [comment]
@ahg-g good point, and I missed that the KEP was still under the wrong directory before it got merged. How can we fix that?

Also, we have an issue tracking the test moves here: https://github.com/kubernetes/kubernetes/issues/87427. @fengzixu, do you have any updates on those moves?

**Comment by ahg-g** [comment]
> @ahg-g good point, and I missed that the KEP was still under the wrong directory before it got merged. How can we fix that?

I would submit a PR to remove the KEP from sig scheduling, and then another one to place it under keps/sig-node

> 
> Also, we have an issue tracking the test moves here: #87427. @fengzixu, do you have any updates on those moves?

File: test/integration/scheduler/taint_test.go
Comment (skilxn-go) [reply test/integration/scheduler/taint_test.go]: Yes, and it's done now

**Comment by fengzixu** [comment]
> @ahg-g good point, and I missed that the KEP was still under the wrong directory before it got merged. How can we fix that?
> 
> Also, we have an issue tracking the test moves here: #87427. @fengzixu, do you have any updates on those moves?

@damemi  Working on it. Submit PR later.

**Comment by damemi** [comment]
@ahg-g once https://github.com/kubernetes/kubernetes/pull/88152 merges, can this be approved as well?

**Comment by ahg-g** [comment]
> @ahg-g once #88152 merges, can this be approved as well?

The KEP PR isn't merged yet: https://github.com/kubernetes/enhancements/pull/1510

**Comment by damemi** [comment]
@ahg-g right, #88152 will merge after the KEP. Is that ok? Trying to establish an order for these PRs to get them merged in time

**Comment by ahg-g** [comment]
> @ahg-g right, #88152 will merge after the KEP. Is that ok? Trying to establish an order for these PRs to get them merged in time

Sounds good as long as we have an understanding with sig node regarding ownership

**Comment by damemi** [comment]
@ahg-g the kep has merged to sig-node, and we are just waiting for approval on https://github.com/kubernetes/kubernetes/pull/88152. Is this ready to approve as well? (We can /hold until the test PR merges)

After this is just the docs change (https://github.com/kubernetes/website/pull/19302)

**Comment by damemi** [comment]
/hold cancel
/sig node
@mattjmcnaughton could you please provide a sig-node signoff for this?

Also, for the other approver sections (https://github.com/kubernetes/kubernetes/pull/87487#issuecomment-578037453), if you could please approve:
/assign @deads2k 
/assign @Huang-Wei 
/assign @k82cn 
/assign @liggitt

**Comment by damemi** [comment]
/assign @dchen1107 
for Node features approver also

**Comment by Huang-Wei** [comment]
/lgtm
/approve

on scheduling side.

**Comment by damemi** [comment]
/hold
For sig-node and api approval

**Comment by damemi** [comment]
Bumping, @smarterclayton will you have time to review this today for code freeze? If not can we file an exception for the enhancement?

**Comment by liggitt** [comment]
/approve

API change is comments only, lgtm

**Comment by liggitt** [comment]
still needs node approval

**Comment by damemi** [comment]
Thanks @liggitt !

@derekwaynecarr, could you approve for sig node?

**Comment by derekwaynecarr** [comment]
changes look good from me for sig-node.

/approve
/lgtm

**Comment by Huang-Wei** [comment]
@skilxn-go can you rebase this PR?

**Comment by mattjmcnaughton** [comment]
> /hold cancel
> /sig node
> @mattjmcnaughton could you please provide a sig-node signoff for this?
> 
> Also, for the other approver sections ([#87487 (comment)](https://github.com/kubernetes/kubernetes/pull/87487#issuecomment-578037453)), if you could please approve:
> /assign @deads2k (kube-controller-manager)
> /assign @Huang-Wei (scheduling)
> /assign @k82cn (nodelifecycle)
> /assign @liggitt (api)
> 
> (there is not a lot of overlap between the different codepath owners in this PR)

Seeing @derekwaynecarr approved on behalf of sig-node. Thanks Derek!

**Comment by skilxn-go** [comment]
@Huang-Wei Sure, I will do it today.

**Comment by skilxn-go** [comment]
Sorry for my delay, I have some issues on cloning my repo from GitHub. I will fix it ASAP.

**Comment by skilxn-go** [comment]
@Huang-Wei It's ok now.

**Review by damemi — APPROVED** [review]
/lgtm
thanks again @skilxn-go

**Pull request merged**
